From 3ca72a84ea5f7d31485cdb60b0d4594db2c61458 Mon Sep 17 00:00:00 2001
From: Mohan Kumar G <quic_mkumarg@quicinc.com>
Date: Mon, 24 Jun 2024 18:28:57 +0530
Subject: [PATCH] hostapd: add puncturing bitmap to frequency parameters

Currently, the puncturing bitmap is not sent from
hostapd to MAC80211. Add punct_bitmap in frequency parameters
and pass the bitmap to MAC80211 via NL80211.

Signed-off-by: Mohan Kumar G <quic_mkumarg@quicinc.com>
---
 src/common/hw_features_common.c | 1 +
 src/drivers/driver.h            | 8 ++++++++
 src/drivers/driver_nl80211.c    | 4 ++++
 3 files changed, 13 insertions(+)

diff --git a/src/common/hw_features_common.c b/src/common/hw_features_common.c
index 13396e4..feec260 100644
--- a/src/common/hw_features_common.c
+++ b/src/common/hw_features_common.c
@@ -514,6 +514,7 @@ int hostapd_set_freq_params(struct hostapd_freq_params *data,
 	data->center_freq2 = 0;
 	data->bandwidth_device = bandwidth_device;
 	data->center_freq_device = center_freq_device;
+	data->punct_bitmap = punct_bitmap;
 	if (oper_chwidth == CONF_OPER_CHWIDTH_80MHZ)
 		data->bandwidth = 80;
 	else if (oper_chwidth == CONF_OPER_CHWIDTH_160MHZ ||
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index 49572fd..7cfa4e5 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -942,6 +942,14 @@ struct hostapd_freq_params {
 	 * one edge of the operating bandwidth.
 	 */
 	int center_freq_device;
+
+	/**
+	 * punct_bitmap - Puncturing bitmap
+	 * Each bit corresponds to a 20 MHz subchannel, lowest bit for the
+	 * channel with the lowest frequency. Bit set to 1 indicates that the
+	 * subchannel is punctured, otherwise active.
+	 */
+	u16 punct_bitmap;
 };
 
 /**
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index ce0320c..afa0dac 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -5127,6 +5127,10 @@ static int nl80211_put_freq_params(struct wpa_driver_nl80211_data *drv,
 		     nla_put_u32(msg, NL80211_ATTR_CENTER_FREQ2,
 				 freq->center_freq2)))
 			return -ENOBUFS;
+
+		if (freq->punct_bitmap &&
+		    nla_put_u16(msg, NL80211_ATTR_RU_PUNCT_BITMAP, freq->punct_bitmap))
+			return -ENOBUFS;
 	} else if (freq->ht_enabled || ((freq->he_enabled || freq->eht_enabled) && is_24ghz)) {
 		enum nl80211_channel_type ct;
 
-- 
2.34.1

